File: custom_sort_string.go
func customSortString(order string, s string) string {
    mp := make([]int, 26)
    var ans []rune
    for _, val := range s {
        mp[val-'a']++
    }

    for _, val := range order {
        for mp[val-'a'] > 0 {
            ans = append(ans, val)
            mp[val-'a']--
        }
    }

    for ch := 0; ch < 26; ch++ {
        for mp[ch] > 0 {
            ans = append(ans, 'a'+rune(ch))
            mp[ch]--
        }
    }
    return string(ans)
}

// Time Complexity : O(n)
// Space Complexity : O(n)

File: custom_sort_string.cpp
class Solution {
public:
    string customSortString(string order, string s) {
        vector<int> mp(26, 0);
        for(int i=0; i < order.size(); i++) {
            mp[order[i]-'a'] = i;
        }

        sort(s.begin(), s.end(), [&](char ch1, char ch2) {
            return mp[ch1-'a'] < mp[ch2-'a'];
        });
        return s;
    }
};

// Time Complexity : O(n*log(n))
// Space Complexity : O(n)